web/app/common/components/log-viewer/log-line/log-line.component.ts

```diff
@@ -0,0 +1,83 @@
+import { DOCUMENT } from '@angular/common';
+import { Component, AfterViewInit, ViewChild, ElementRef, Input, Inject, ViewEncapsulation } from '@angular/core';
+
+const ANSI_PATTERN = /\u001b\[([0-9;]+)?m/;
+const ANSI_RESET_CODE = '0';
+
+type AnsiCode = string;
+type AnsiTuple = [AnsiCode, string];
+
+export interface LogLine {
+  timestamp: number;
+  message: string;
+  level: string;
+  status: number;
+}
+
+@Component({
+  selector: 'fci-log-line',
+  templateUrl: './log-line.component.html',
+  styleUrls: ['./log-line.component.scss'],
+  encapsulation: ViewEncapsulation.None
+})
+
+export class LogLineComponent implements AfterViewInit {
+  @Input() log: LogLine;
+  @ViewChild('logLine', { read: ElementRef }) logLineEl: ElementRef;
+
+  constructor(@Inject(DOCUMENT) private readonly document: any) {}
+
+  ngAfterViewInit() {
+    const parentEl = this.logLineEl.nativeElement;
+    let currentSpanEl = parentEl;
+    const stack = this.tokenize(this.log.message);
+
+    for (const tuple of stack) {
+      const [code, text] = tuple;
```

Comment:
you can actually merge the previous two lines

```
for (const [ansiCode, text] of stack) { ... }
```

Reply:
I tried... Typescript complained at me.